lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,310 @@
+};
+
+/**
+ * Get authorization code lifetime.
+ */
+
+AuthorizeHandler.prototype.getAuthorizationCodeLifetime = function() {
+  var expires = new Date();
+
+  expires.setSeconds(expires.getSeconds() + this.authorizationCodeLifetime);
+
+  return expires;
+};
+
+/**
+ * Get the client from the model.
+ */
+
+AuthorizeHandler.prototype.getClient = function(request) {
+  var clientId = request.body.client_id || request.query.client_id;
+
+  if (!clientId) {
+    throw new InvalidRequestError('Missing parameter: `client_id`');
+  }
+
+  if (!is.vschar(clientId)) {
+    throw new InvalidRequestError('Invalid parameter: `client_id`');
+  }
+
+  var redirectUri = request.body.redirect_uri || request.query.redirect_uri;
+
+  if (redirectUri && !is.uri(redirectUri)) {
+    throw new InvalidRequestError('Invalid request: `redirect_uri` is not a valid URI');
+  }
+
+  return Promise.try(this.model.getClient, clientId)
+    .then(function(client) {
+      if (!client) {
+        throw new InvalidClientError('Invalid client: client credentials are invalid');
+      }
+
+      if (!client.grants) {
+        throw new InvalidClientError('Invalid client: missing client `grants`');
+      }
+
+      if (!_.contains(client.grants, 'authorization_code')) {
+        throw new UnauthorizedClientError('Unauthorized client: `grant_type` is invalid');
+      }
+
+      if (!client.redirectUri) {
+        throw new InvalidClientError('Invalid client: missing client `redirectUri`');
+      }
+
+      if (redirectUri && client.redirectUri !== redirectUri) {
+        throw new InvalidClientError('Invalid client: `redirect_uri` does not match client value');
+      }
+
+      return client;
+    });
+};
+
+/**
+ * Get scope from the request.
+ */
+
+AuthorizeHandler.prototype.getScope = function(request) {
+  var scope = request.body.scope || request.query.scope;
+
+  if (!is.nqschar(scope)) {
+    throw new InvalidScopeError('Invalid parameter: `scope`');
+  }
+
+  return scope;
+};
+
+/**
+ * Get state from the request.
+ */
+
+AuthorizeHandler.prototype.getState = function(request) {
+  var state = request.body.state || request.query.state;
+
+  if (!state) {
+    throw new InvalidRequestError('Missing parameter: `state`');
+  }
+
+  if (!is.vschar(state)) {
+    throw new InvalidRequestError('Invalid parameter: `state`');
+  }
+
+  return state;
+};
+
+/**
+ * Get user by calling the authenticate middleware.
+ */
+
+AuthorizeHandler.prototype.getUser = function(request, response) {
+  return this.authenticateHandler.handle(request, response).then(function(token) {
+    return token.user;
+  });
+};
+
+/**
+ * Save authorization code.
+ */
+
+AuthorizeHandler.prototype.saveAuthorizationCode = function(authorizationCode, expiresAt, scope, client, user) {
+  var code = {
+    authorizationCode: authorizationCode,
+    expiresAt: expiresAt,
+    scope: scope
+  };
+
+  return Promise.try(this.model.saveAuthorizationCode, [code, client, user]);
```

Comment:
Sounds good and makes sense given your [previous comment](https://github.com/thomseddon/node-oauth2-server/pull/203#discussion_r41379580).

:+1: